wfe2/wfe.go

```diff
@@ -2227,12 +2275,60 @@ func (wfe *WebFrontEndImpl) NewOrder(
 
 	logEvent.DNSNames = names
 
+	var replaces string
+	var limitsExempt bool
+	if newOrderRequest.Replaces != "" {
+		// Implementing draft-ietf-acme-ari-02: For a new order to be considered
+		// a replacement for an existing certificate, it:
+		//   1. MUST NOT have been replaced by another finalized order,
+		//   2. MUST be associated with the same ACME account as its predecessor, and
+		//   3. MUST have at least one identifier in common with its predecessor.
+		certID, prob, err := parseCertID(newOrderRequest.Replaces, wfe.issuerCertificates)
+		if err != nil {
+			wfe.sendError(response, logEvent, prob, err)
+			return
+		}
+
+		exists, err := wfe.sa.ReplacementOrderExists(ctx, &sapb.Serial{Serial: certID.Serial()})
+		if err != nil {
+			wfe.sendError(response, logEvent, web.ProblemDetailsForError(err, "Failed to check replacement status of existing certificate"), err)
+			return
+		}
+		if exists.Exists {
+			wfe.sendError(response, logEvent, probs.Malformed("Cannot indicate an order replaces a certificate which already has a replacement order"), nil)
```

Comment:
It seems like this should be a non-error; it just means that this request doesn't get a rate limit exemption. The client probably isn't keeping perfect track of which certificates have already been "replaced", since it's a complicated thing (names added, names removed, maybe a certificate was split). So it's okay to specify a `replaces` field that points to a certificate that has already been replaced.

Reply:
Talked this over live. @beautifulentropy made the good points that a client that is doing incorrect recordkeeping for certificate replacement should receive feedback about that fact, and the most straightforward way to give that feedback is by returning an error. Also, since `replaces` will be used for ordinary certificate replacement, not just during an incident, there are ample opportunities to detect that incorrect recordkeeping during normal operations.

Note that there are some examples of "incorrect recordkeeping" that are unavoidable. For instance a client might crash after requesting a certificate but before recording it. The next request would also specify a `replaces` field, and would get an error (certificate xyz has already been replaced).

IMO the correct thing for ACME clients to implement is: try once with `replaces`; if it fails, try again without `replaces`. To enable that, it would be useful to specify a Problem Type specifically for "certificate has already been replaced", with the semantics that the client should retry without the `replaces` field (and notify the operator?). That can be done as a followup update to the code and the spec.

So the code is fine as it stands, but the error should be more user-centric (similar to the tweak I proposed at https://github.com/letsencrypt/boulder/pull/7298#discussion_r1498410138).

Reply:
https://github.com/aarongable/draft-acme-ari/issues/56 and https://github.com/letsencrypt/boulder/issues/7338 have been created.